### PR TITLE
[Discover] Check if user has any access to any registered resource

### PR DESCRIPTION
**Pull request by kimlisa**
#### Description
part of https://github.com/gravitational/teleport/issues/13997

- adds an endpoint for web apiserver to check if user has any access to a registered resource
- added the new connection diagnostic access perms to UI user context
- tweaked list resources to not always require fake pagination

### CODE REVIEW

File: lib/web/ui/usercontext.go [hunk 1]
```diff
@@ -83,6 +83,8 @@ type userACL struct {
 	AccessRequests access `json:"accessRequests"`
 	// Billing defines access to billing information.
 	Billing access `json:"billing"`
+	// ConnectionDiagnostic defines access to connection diagnostics.
+	ConnectionDiagnostic access `json:"connectionDiagnostic"`
```
Comment (marcoandredinis) [inline lib/web/ui/usercontext.go:87]: Is this change related with the check we are doing?

File: lib/web/resources_test.go [hunk 2]
```diff
@@ -375,7 +375,6 @@ func TestListResources(t *testing.T) {
 			expected: proto.ListResourcesRequest{
 				ResourceType:        types.KindNode,
 				Limit:               defaults.MaxIterationLimit,
-				NeedTotalCount:      true,
```
Comment (marcoandredinis) [inline lib/web/resources_test.go:378]: Why are we removing this field?

File: lib/web/apiserver_test.go [hunk 3]
```diff
@@ -862,8 +862,10 @@ func TestClusterNodesGet(t *testing.T) {
 
 	endpoint := pack.clt.Endpoint("webapi", "sites", clusterName, "nodes")
 
+	query := url.Values{"sort": []string{"name"}}
```
Comment (marcoandredinis) [inline lib/web/apiserver_test.go:865]: Not requiring things to be sorted may reduce database stress
Also, for this particular check seems to be unnecessary (we only care if exists one or none)

File: lib/web/apiserver.go [hunk 4]
```diff
@@ -534,6 +534,7 @@ func (h *Handler) bindDefaultEndpoints(challengeLimiter *limiter.RateLimiter) {
 	h.GET("/scripts/:token/install-app.sh", httplib.MakeHandler(h.getAppJoinScriptHandle))
 	// web context
 	h.GET("/webapi/sites/:site/context", h.WithClusterAuth(h.getUserContext))
+	h.GET("/webapi/sites/:site/resources/check", h.WithClusterAuth(h.checkAccessToRegisteredResource))
```
Comment (marcoandredinis) [inline lib/web/apiserver.go:537]: :nail_care: 

`check` seems too generic but I don't have a better suggestion
Sorry, naming is hard :bow:

**Review by marcoandredinis — APPROVED** [review]
Overall looks good :+1: 
Just left some question

File: lib/web/ui/usercontext.go (hunk 1 above)
Comment (kimlisa) [reply lib/web/ui/usercontext.go:87]: i didn't know how literally to take this question :sweat_smile: so i provided both

yes, it's related in that it's required for discovery, so we can pre-check that users have all the necessary perms before allowing user to start the discovery flow

no, in that it's not required for the `does user has at least 1 resource` check, i just sneaked it in b/c it was tiny add in (maybe i should've have?)

File: lib/web/resources_test.go (hunk 2 above)
Comment (kimlisa) [reply lib/web/resources_test.go:378]: so before, i set this field to always be true because i thought the list resources request coming through the web api layer will always require it (bad decision on my part), but this field isn't actually necessary because our frontend calls with the query `sort` always set, which will also return the total count.

File: lib/web/apiserver_test.go (hunk 3 above)
Comment (kimlisa) [reply lib/web/apiserver_test.go:865]: it's related to this comment https://github.com/gravitational/teleport/pull/15637#discussion_r952105198, the tests that include the `sort` query params, has a test that tested for the `totalCount` which can only be returned with calls to ListResources that triggers fake pagination (sort being one of them)

File: lib/web/apiserver.go (hunk 4 above)
Comment (kimlisa) [reply lib/web/apiserver.go:537]: i was actually hoping someone had a better name but alas :sweat_smile:

File: lib/web/ui/usercontext.go (hunk 1 above)
Comment (marcoandredinis) [reply lib/web/ui/usercontext.go:87]: Got it :+1: 
Thank you

File: lib/web/resources.go [hunk 5]
```diff
@@ -340,6 +378,12 @@ type listResourcesGetResponse struct {
 	TotalCount int `json:"totalCount"`
 }
 
+type checkAccessToRegisteredResourceResponse struct {
+	// HasResource is a flag to indicate if user has any access
+	// to a registered resource or not.
+	HasResource bool `json:"hasResource"`
```
[the remark this answers is not in the document]
Comment (kimlisa) [reply lib/web/resources.go:384]: i don't think this is necessary, if this endpoint returns true we can use the perms acquired by fetching user context to determine which first tab the user has full acess to

**Review by r0mant — APPROVED** [review]
